core/src/main/java/org/apache/iceberg/avro/TypeToSchema.java

```diff
@@ -214,4 +220,46 @@ public Schema primitive(Type.PrimitiveType primitive) {
 
     return primitiveSchema;
   }
+
+  private static String sanitize(String name) {
+    int length = name.length();
+    StringBuilder sb = new StringBuilder(name.length());
+    char first = name.charAt(0);
+    if (!(Character.isLetter(first) || first == '_')) {
+      sb.append(sanitize(first));
+    } else {
+      sb.append(first);
+    }
+    for (int i = 1; i < length; i++) {
+      char character = name.charAt(i);
+      if (!(Character.isLetterOrDigit(character) || character == '_')) {
+        sb.append(sanitize(character));
+      } else {
```

Comment:
What about an additional `else if` here that checks for numbers? That way, the `9x` test case could be `_9x` instead of `_x39x`. No need to use the hex encoding if it is a valid character.

Reply:
Instead of adding another else if here, I can update the `sanitize` method to handle numbers like you proposed.